package/MDAnalysis/analysis/Aggregation_size.py

```diff
@@ -0,0 +1,370 @@
+            for i in range(j+1, self._molnum):
+                d[j+1][ii] = dists[i][j]
+                ii += 1
+
+    def specific_atoms(self, d):
+        """
+        This function calculates the distance between the selected atom
+        of two molecules.
+        """
+        if (not(atom_1 or atom_2)):
+            raise ValueError("enter valid atom_1 and atom_2")
+        self._cut_off = self._cut_off or 8.5
+        residue1 = u.select_atoms(f"name {atom_1}").positions
+        residue2 = u.select_atoms(f"name {atom_2}").positions
+        dists = distances.distance_array(
+            residue1, residue2, box=u.universe.dimensions)
+        for j in range(self._molnum):
+            ii = 0
+            for i in range(j+1, self._molnum):
+                d[j+1][ii] = dists[i][j]
+                ii += 1
+
+    def _prepare(self):
+        """
+        Create array of zeroes as a placeholder for differnet parameters.
+        """
+        self.results = np.zeros((self.n_frames, 5))
+        self.d = [0] * self._molnum
+        i = 0
+        for it in reversed(range(1, self._molnum+1)):
+            self.d[i] = np.zeros(it * 1).reshape(it, 1)
+            i += 1
+        self.size_num = [0] * self.n_frames
+
+    def aggregation(self, k, tss, counter, m, cut_off, connected):
+        """
+        This is a subfunction that enables the code to find all
+        connected molecules in a cluster.
+        k is a connected molecule to the main molecule in the first level.
+        self._frame_index is a time frame.
+        counter is a step.
+        m provides a suitable off_set to match the number of molecules.
+        Limit is the cut_off provided by the user.
+        """
+        bool_sector = self.d[k] < cut_off
+        Coordinates = self.d[k][bool_sector]
+        # This "if" will check if there is any uninvestigated molecule left
+        if len(Coordinates) != 0:
+            index1 = np.where(bool_sector)
+            listOfCoordinates = list(zip(index1[0], index1[1]))
+            connect = ([j[0] for j in listOfCoordinates]
+                       + connected[m] + 1)
+            Connected_NLevel = np.array(connect)
+        else:
+            Connected_NLevel = []
+        return Connected_NLevel
+
+    def _single_frame(self):
+        """
+        1- Import the needed module
+        2- Calculate the distance between molecules based on the intended
+        criteria
+        3- Find the aggregates by checking only one time each using the
+        "aggregate" subfunction (to be efficient)
+        4- Check and merge two aggregates as one aggregate if the distance
+        between any element of them is less than cut_off
+        5- Eliminate monomers (optional)
+        6- measure average gyration radius (optional)
+        7- measure the number of aggregates with different size
+        """
+        aggregate = [0] * self._molnum
+        if self._type == "Closest":
+            self.closest_atom(self.d)
```

Comment:
Relevant to @orbeckst's comment above, I would have an umbrella function `d = self._get_distances()` which you assign in `__init__`:

```python
def __init__(self, method="closest", ...):
    if method == "closest":
        self._get_distances = self._get_distances_by_closest_atoms
    ... # etc
```

Renaming your other functions with action words makes it clearer what they do, and returning the `d` distance array means you don't overwrite it every frame.

If you want to save your distances, you can prepare a placeholder array in `_prepare`:

```
self._distances = np.zeros((self.n_frames, self.n_residues, self.n_residues))
```

or if you would like to flatten it to save space:

```
n_triangle = self.n_residues * (self.n_residues-1) / 2
self._distances = np.zeros((self.n_frames, n_triangle))
```

And in `single_frame`, you could update `self._distances[self._frame_index] = d`.

Reply:
Thank you, I used the return as @orbeckst mentioned. I do not need to save distance of all frame that is why I am replacing them every frame.  I also changed the name of parameter to make them more readable and understandable. you will see that in the new version of code.